[CoreIPC] [GPU] WebCore::SVGFilter expression/effects members are not validated

jon-b-apple opened the pull request:
#### 45047bcfe94e948beb0f3a2ca6cebf03fa2a118f
<pre>
[CoreIPC] [GPU] WebCore::SVGFilter expression/effects members are not validated
<a href="https://bugs.webkit.org/show_bug.cgi?id=289947">https://bugs.webkit.org/show_bug.cgi?id=289947</a>
<a href="https://rdar.apple.com/142968121">rdar://142968121</a>

Reviewed by Said Abou-Hallawa.

Introduce validation at the IPC boundary to ensure SVGFilter expression indicies are in-bounds of the provided effects.

* LayoutTests/ipc/invalid-svgfilter-expression-crash-expected.txt: Added.
* LayoutTests/ipc/invalid-svgfilter-expression-crash.html: Added.
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::isValidSVGFilterExpression):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/292483@main">https://commits.webkit.org/292483@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21a20b74aa898c659e3afdd735702552abb85a7a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95993 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15607 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5553 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101055 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46502 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98038 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15902 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24039 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73182 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30408 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98996 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11898 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86703 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53518 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11642 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4455 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45837 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81792 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4556 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103081 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23060 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16797 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82222 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23311 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82724 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81595 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20514 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26185 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3623 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/16399 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23023 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28178 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22682 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26162 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24423 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -326,6 +327,16 @@ RefPtr<FilterImage> SVGFilter::apply(FilterImage* sourceImage, FilterResults& re
     return stack.takeLast();
 }
 
+bool SVGFilter::validateEffects(const SVGFilterExpression& expression, const FilterEffectVector& effects)
```
I would suggest using a name which starts with the verb **is** something like `isValidSVGFilterExpression()` or something similar.

shallawa commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -296,6 +296,7 @@ RefPtr<FilterImage> SVGFilter::apply(FilterImage* sourceImage, FilterResults& re
     FilterImageVector stack;
 
     for (auto& term : m_expression) {
+        ASSERT(term.index < m_effects.size());
```
I do not think we should repeat this `ASSERT` every time we access `m_effects`. This is what I think 

1. The second `SVGFilter::create()` and the second `SVGFilter:: SVGFilter()` constructor are not used. So please delete them. 
2. In the remaining two creates add just one `ASSERT`. For example the first `SVGFilter::create()` can have this `ASSERT`:

```
auto& expression = std::get<SVGFilterExpression>(*result);
auto& effects = std::get<FilterEffectVector>(*result);
ASSERT(validateEffects(expression, effects));
```

jon-b-apple replied:
Ok, I can remove the unused SVGFilter create and constructor call if it's not too far outside the scope of this fix.

The call to the validation function happens in the IPC layer, by the time create is called we have already succeeded in validating the expression. The reason for adding ASSERTs on each index into expression is because I can't confirm that there aren't any non-IPC paths to SVGFilter::apply and the other accesses, and these won't be validated. 

I reasoned that we could use ASSERT rather than RELEASE_ASSERT based on the WTF::Vector class always doing bounds checks in all configurations - given that ASSERT only runs in debug builds, is there an issue keeping the ASSERTs near to the accesses? My understanding is that these exist solely to make it more clear what the issue was if we were to somehow call this with a crashing input in debug builds only.

jon-b-apple replied:
I removed the unused constructor and create call, but decided to leave the ASSERTs where they are based on the reasoning above. I feel that having asserts for conditions close to the code that relies on that assertion is the clearest, and while there's some repetition in the code, there's no runtime cost for this in release builds. Even if we moved the ASSERTs to both of the remaining create calls, there would only be 1 fewer ASSERT in the file, and we'd have to call the validation function again rather than just checking each index as we process them.

shallawa replied:
I removed the unused code in https://commits.webkit.org/292339@main.

Since Vector does the bounds checks in all its operations, 
1. What is a new `ASSERT` going to add here? 
2. Why do not we change all the calls in all the sources to `ASSERT(term.index < vector.size());` before we access `vector[index]`?

The benefit of adding the validation in the `SVGFilter::create()` function is we make sure the validation was done at the IPC layer. If the validation was deleted by mistake from the IPC layer, the `ASSERT(validateEffects(expression, effects));` will fire early enough to see its cause clearly.

jon-b-apple replied:
The ASSERTs in their current position are purely informational, to allow debug builds to crash with a more informative error and stack, rather than ending up in WTF's Vector OverflowHandler. They're there to catch any possible paths through the code that might not enter via IPC, and therefore not be validated - both apply and createFilterStyles are marked as WEBCORE_EXPORT in the WebCore::Filter base class, and this makes it unclear to me that all possible paths to the functions are preceded by an IPC receive and validate.

In early versions of the patch these were ASSERT_WITH_SECURITY_IMPLICATIONs, but following the advice of the WK style checker I made them regular ASSERTs. I chose not to use RELEASE_ASSERTs as I had originally suggested to avoid duplicating the validation already done by WTF::Vector. You're correct that this logic could be applied to any place we have a controllable index being used in an operator[]-like call, it's a tradeoff between terseness and clarity. We can remove the ASSERTs if you think that is a better tradeoff here, but it seems like there's precedent for having explicit assumptions about what we expect in the code - for example the `ASSERT(stack.size() == 1)` at the end of SVGFilter::apply isn't strictly needed since takeLast calls at which checks the index is in bounds, but the assertion makes it clear what we expect to be true before we hit any crashes should the invariant not be satisfied.

Another reason to keep the ASSERT calls next to the index is that it's possible (although doesn't seem to happen in the current code) that the expressions are changed after the call to SVGFilter::create. The setExpression method on SVGFilter could update the expression after a call to create, and while it's private it might be exposed via other mechanisms like friend declarations. In my opinion, it's much more likely that someone would delete the ASSERT from the create call than from the IDL Validator attribute, since in the create call it's not necessarily clear why the validation is relevant. With ASSERTs directly before the use, it's clear why we are making the assertion even though non-debug builds will still crash in the OverflowHandler and need to walk back up the stack to understand why the crash happened.

shallawa replied:
`ASSERT(stack.size() == 1)` at the end of `SVGFilter::apply()` ensures that the postfix expression is evaluated correctly. If `SVGFilter::apply()` does it correctly, the stack has to have one and only one element at the end. It is a bug if the stack has 2 elements for example even though `takeLast()` will not complain.

jon-b-apple replied:
In release builds the assert does nothing, so this invariant is actually not enforced here. The point was to highlight that ASSERTs commonly exist next to the code that relies upon them. They exist to check additional invariants in build configurations that compile them in (debug only to my knowledge), or to make conditions that would otherwise be memory access violations more clear by crashing with an informative message. As a code reviewer, they also communicate what the developer expects to be true at a given point in a function, which is another reason to have them as close as possible to code that relies on them. 

Having an ASSERT far away from the code that relies on the condition being asserted makes the code less clear in my opinion - it makes it more likely the assertion will be removed as it's not clear why it's there in the first place, and it opens the possibility that the condition being checked can change between the check and the use.

jon-b-apple replied:
I removed the ASSERTs to unblock landing this patch. To reiterate, these were not performing any part of the validation or enforcing any security guarantees, they were simply there to help clarify the reason for any crashes when indexing into m_effects in debug builds only.

shallawa commented on LayoutTests/ipc/invalid-svgfilter-expression-crash.html:
```diff
@@ -0,0 +1,95 @@
+                        primitiveUnits: 0,
+                        expression: {
+                            alias: [
+                                {index: 1337, level: 0, geometry: 0}    // Adds an expression with an OOB index
```
Why does not this test crash with this OOB index? Is this because of adding the `Validator` in WebCoreArgumentCoders.serialization.in?

jon-b-apple replied:
Yes, the validation added will detect the OOB index and fail deserialisation. This will result in a TypeError being thrown by the IPC Testing API. Without the validation, the message makes it through to the dereference which will crash the GPU process.

shallawa commented on LayoutTests/ipc/invalid-svgfilter-expression-crash.html:
```diff
@@ -0,0 +1,95 @@
+        });
+    } catch(err) {
+        if(!(err instanceof TypeError)) {
+            console.log("Expected TypeError, got " + err);
```
Is it expected that WebKit does not crash and no exception was thrown? So how can we confirm we validated the expression in this test?

jon-b-apple replied:
With the validation in place, it's expected that we get a TypeError indicating deserialisation failed. I wanted to log (and fail the test) if we get any other kind of error, as that might indicate the message wasn't sent and so the test could be non-functional. It's not perfect, since I think a change in message structure would also return the same TypeError. I'd be happy to hear any recommendations you might have to better detect the difference between and invalid message structure and a validation failure.

jon-b-apple replied:
As we introduce more validator tests I suspect this might become more of an issue, so I filed an enhancement radar (rdar://147337600) for the IPC testing API to be able to report validation errors separately to deserialisation errors.

jon-b-apple replied:
In the short-term, I added some additional comments to explain why I am checking for TypeError here, and made the log message more clear that anything else should be considered a test failure rather than errand logging left in from debugging. Once the enhancement radar is fixed, we should update the test to make sure the error came specifically from a validation error.

shallawa commented on LayoutTests/ipc/invalid-svgfilter-expression-crash.html:
```diff
@@ -0,0 +1,97 @@
+    } catch(err) {
+        // We expect validation to fail and return a TypeError to us. If we get any other kind of error,
+        // log this so we can address the issue as this may indicate the test is non-functional
+        if(!(err instanceof TypeError)) {
+            console.log("Test failed: expected TypeError, got " + err);
```
Can't we capture this error in the expected file? It is the only evidence that the validation has run on the SVGFilter.

jon-b-apple replied:
In the current state, I don't think a TypeError conclusively says the validation ran, just that the message was rejected due to it failing somewhere in the deserialisation process. This might be because of a validation failure, but it could also be that the message structure changed and the decoder was marked as invalid at some point during decoding.

When the enhancement I filed is implemented we should definitely find a way to assert that the validation was done and the message was rejected for that reason specifically. Matching the output to something in the expected file would be a good way to do this, but right now I don't think it buys us anything. My checking the TypeError here is just extra caution in case there are any other kinds of JS error that we might get propagated back to coreipc.js, to allow us to go and investigate.

shallawa replied:
At least a TypeError will say there is something wrong with your test case. What is  the enhancement that you filed? Can you include a link?

jon-b-apple replied:
We will get a TypeError both in the case where validation rejects the message (test should pass) AND in the case where e.g. the message structure has changed (test should fail). Currently I believe there is no way to differentiate between them. The expected behaviour once validation is in place is a TypeError, any other outcome at this point is unexpected, and so I log a message which will cause the test to fail and allow us to investigate why.

The enhancement radar is rdar://147337600, which when implemented should allow us to differentiate between a validation error (expected for the test to pass) and any other kind of deserialisation error (indicating a problem with the test, upon which we will likely log this and cause the test to fail). Whether we log on success and add it to the expected output or log on failure and cause the test to fail is our choice, but currently I don't believe we can do any better than what I have written here. Again, this check is out of an abundance of caution, I'm not currently sure we can even get a different kind of JS error. This check guards against that possibility (now or in the future), but until we have a way to differentiate validation errors from other kinds of deserialisation errors I believe this is the best we can do.

jon-b-apple replied:
I added comments to clarify the expected and unexpected results of sending the message, as well as highlight the limited ability we have currently to detect validation failures over other kinds of deserialisation errors and linked the enhancement radar.

jon-b-apple committed:
[CoreIPC] [GPU] WebCore::SVGFilter expression/effects members are not validated
https://bugs.webkit.org/show_bug.cgi?id=289947
rdar://142968121

Reviewed by Said Abou-Hallawa.

Introduce validation at the IPC boundary to ensure SVGFilter expression indicies are in-bounds of the provided effects.

* LayoutTests/ipc/invalid-svgfilter-expression-crash-expected.txt: Added.
* LayoutTests/ipc/invalid-svgfilter-expression-crash.html: Added.
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::isValidSVGFilterExpression):
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/292483@main
